packages/tools/pluggable-widgets-tools/configs/rollup-plugin-collect-native-deps.js

```diff
@@ -0,0 +1,110 @@
+import { basename, dirname, extname, join } from "path";
+import { copy, readJson, writeJson } from "fs-extra";
+import { promises } from "fs";
+import { rm } from "shelljs";
+
+export function collectNativeDependencies({ outputDir, externals, widgetName }) {
+    let firstRun = true;
+    const nativeDependencies = [];
+    const nodeModulesPath = join(outputDir, "node_modules");
+    return {
+        name: "collect-native-deps",
+        async resolveId(source) {
+            if (source.startsWith(".")) {
+                return null;
+            }
+
+            try {
+                const packageFilePath = require.resolve(`${source}/package.json`);
+                const packageDir = dirname(packageFilePath);
+
+                if (await hasNativeCode(packageDir)) {
+                    if (!nativeDependencies.some(x => x.name === source)) {
+                        nativeDependencies.push({ name: source, dir: packageDir });
+                    }
+                    return { id: source, external: true };
+                }
+                return null;
+            } catch (e) {
+                return null;
+            }
+        },
+        async buildStart() {
+            rm("-rf", nodeModulesPath);
+            if (!firstRun) {
+                nativeDependencies.length = 0;
+            }
+            firstRun = false;
+        },
+        async writeBundle() {
+            await Promise.all(
+                nativeDependencies.map(async dependency => {
+                    await copyJsModule(dependency.dir, join(nodeModulesPath, dependency.name));
+                    for (const transitiveDependency of await getTransitiveDependencies(dependency.name, externals)) {
+                        await copyJsModule(
+                            dirname(require.resolve(`${transitiveDependency}/package.json`)),
+                            join(nodeModulesPath, dependency.name, "node_modules", transitiveDependency)
+                        );
+                    }
+                })
+            );
+            await writeNativeDependenciesJson({ nativeDependencies, outputDir, widgetName });
```

Comment:
I think we should also catch any errors thrown here

Reply:
Done